### ISSUE TITLE
Add a accessible object pool for different data structures

### ISSUE THREAD

**Issue by andersfylling** [andersfylling/disgord#89]
The flyweight pattern can greatly reduce the number of mem allocs. However, it requires us to keep track of ownership so we don't create race conditions. 

I don't know exactly how this can be done. But my current idea for now is to provide two new DisGord functions to set ownership of a discord struct:
 - `disgord.TakeOwnership(d discordStruct)`
 - `disgord.ReleaseOwnership(d discordStruct)`

Then initialisation of structs would be through a factory/constructor that is aware of available/reusable resources of the same type.

These would only be needed if the developer uses the structs outside the handlers, eg. sending structs into a new go routine or a local cache. As long as they don't utilise the object after all their handlers are run, it's no issue for us to zero initialise the struct and re-use it. I believe this is easiest, we could do a hash and compare it to the next event. But I don't expect any gains from that as we rarely would get duplicate objects.

Also, I have no idea how to do this for event channels, or long running processes/services that extracts content from the cache or rest methods or another way of extract Discord data.

Any ideas and experience is greatly appreciated.

Edit: can build constraints be used to activate this behavior? Making the developer responsible for proper use. While others won't have to worry about it, at all.

Edit2: Just a object pool that takes objects implementing the Reseter interface is sufficient.

**Comment by andersfylling** [comment]
By default a private pool will be used for websocket events. This forcers developers to not share the event object(s) in the handlers.

A way around this would be to call .DeepCopy() on incoming events in the handler, or use a build flag "disgord_no_evtpool". Which should be enough.

**Comment by andersfylling** [comment]
Implementing sync.Pool for MessageCreate events showed to be slower than allocating the objects when needed.

With pool:
```
➜ go test -bench BenchmarkClient_On -benchtime 1m -benchmem
goos: linux
goarch: amd64
pkg: github.com/andersfylling/disgord
BenchmarkClient_On-8       10000000          8826 ns/op        1876 B/op          54 allocs/op
PASS
ok      github.com/andersfylling/disgord    96.073s
```

Without pool:
```
➜ go test -bench BenchmarkClient_On -benchtime 1m -benchmem
goos: linux
goarch: amd64
pkg: github.com/andersfylling/disgord
BenchmarkClient_On-8       10000000          8771 ns/op        1904 B/op          55 allocs/op
PASS
ok      github.com/andersfylling/disgord    96.666s
```

With these findings, a resource pool will not be implemented for v0.10. It might however be revised later on. Closing for now.

That being said, it can be interesting to see how a non-blocking solution would work. Eg. https://timharris.uk/papers/2001-disc.pdf
